File: MVVM_ROOM/app/src/main/java/com/tyagiabhinav/loremipsum/viewmodel/PostsViewModel.java
package com.tyagiabhinav.loremipsum.viewmodel;

import android.app.Application;

import com.tyagiabhinav.loremipsum.model.PostsRepository;
import com.tyagiabhinav.loremipsum.model.db.Posts;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import androidx.annotation.NonNull;
import androidx.databinding.ObservableField;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import dagger.hilt.android.lifecycle.HiltViewModel;

@HiltViewModel
public class PostsViewModel extends AndroidViewModel {

    PostsRepository postsRepository;
    public final ObservableField<Boolean> isLoading = new ObservableField<>(false);
    private final MutableLiveData<List<Posts>> allPosts;

    @Inject
    public PostsViewModel(@NonNull Application application, PostsRepository repo) {
        super(application);
        this.allPosts = new MutableLiveData<>();
        this.postsRepository = repo;
    }

    public LiveData<List<Posts>> getAllPosts() {
        isLoading.set(true);
        postsRepository.getAllPosts(new PostsRepository.RepoResult() {
            @Override
            public void onSuccess(List<Posts> postsList) {
                isLoading.set(false);
                allPosts.postValue(postsList);
            }

            @Override
            public void onError(String message, Exception exception) {
                isLoading.set(false);
                allPosts.postValue(new ArrayList<>());
            }
        });
        return allPosts;
    }


    public void setVisibilityState(int pos, boolean isVisible){
        allPosts.getValue().get(pos).setDescVisible(isVisible);
    }
//
//    public boolean isVisible(int pos){
//        return allPosts.getValue().get(pos).isDescVisible();
//    }

    /*
    insert, update, delete also goes here if data is being changed by the user
     */

}

File: MVVM_ROOM/app/src/main/java/com/tyagiabhinav/loremipsum/viewmodel/PostsViewModel.kt
package com.tyagiabhinav.loremipsum.viewmodel

import android.app.Application
import androidx.databinding.ObservableField
import androidx.lifecycle.*
import com.tyagiabhinav.loremipsum.model.Repository
import com.tyagiabhinav.loremipsum.model.dao.Post
import com.tyagiabhinav.loremipsum.model.dao.ResultState
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.Dispatchers.IO
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class PostsViewModel
@Inject constructor(
        private val app: Application,
        private val repo: Repository
) : AndroidViewModel(app) {

    val dataLoading = ObservableField(false)

    val result: LiveData<ResultState<List<Post>>> = liveData(IO, 5000) {
        dataLoading.set(true)
        try {
            val res = repo.getPosts()
            if (res.isNotEmpty()) {
                emit(ResultState.Success(res))
            } else {
                emit(ResultState.Error(Exception()))
            }
        } catch (ioException: Exception) {
            emit(ResultState.Error(Exception()))
        } finally {
            dataLoading.set(false)
        }
    }
}